nntrainer/layers/embedding.cpp

```diff
@@ -133,10 +135,38 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void EmbeddingLayer::calcDerivative() {
-  // NYI
+  return; // intended
 }
+
 void EmbeddingLayer::calcGradient() {
-  // NYI
+  Tensor &djdw =
+    weightAt(static_cast<int>(EmbeddingParams::weight)).getGradientRef();
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  djdw.setZero();
+
+  // TODO:
+  // This is to calculate gradient with current implementation of optimizer.
+  // In order to accelerate, we need to better way like using index to weight.
+
+  for (unsigned int b = 0; b < input_.batch(); ++b) {
+    float *in_data = input_.getAddress(b * input_.getDim().getFeatureLen());
+
+    for (unsigned int i = 0; i < in_length; ++i) {
+      if (in_data[i] - 1 < 0)
```

Comment:
This is not needed. This check was already verified in `forwarding`.

Reply:
It is needed. it is for the skipping of calculation.

Reply:
Can you please add in the comments the scenario where in_data[i] is less than 1?
Also, can you write the semantics of in_data - that it must be an unsigned integer etc.

Reply:
Like I describe earlier, it is padding.